In [1]:
#importing library
import os
import json
import pandas as pd
import glob
from pandas.io.json import json_normalize

#Display all columns for excel/csv files
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

#Ignore warnings
import warnings
warnings.filterwarnings("ignore")

# Reading each Folder to extract the JSON files

## Optimization Data 1

### Loan Data-1

In [2]:
# Read multiple json files from a directory and convert them to pandas dataframe
# Add file names as a new column in DataFrame
path_to_json = '/Users/apurwaprasad/Desktop/BlackKnight_data_challenge/Mortgage_Pooling_v1/OptimizationData1/LoanData1/'

data_list = []
json_file_names = []
for file_name in [file for file in os.listdir(path_to_json) if file.endswith('.json')]:
    with open(path_to_json + file_name) as json_file:
        image_name = os.path.basename(file_name)
        data_name = os.path.splitext(image_name)[0]
        data = json.load(json_file)
        data_list.append(data)
        json_file_names.append(data_name)
        
LoanData1 = pd.DataFrame.from_dict(json_normalize(data_list))
LoanData1['file_name'] = json_file_names

LoanData1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 309 entries, 0 to 308
Data columns (total 17 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   cID          309 non-null    object 
 1   AmortTerm    309 non-null    int64  
 2   Amount       309 non-null    int64  
 3   cCode        309 non-null    object 
 4   Channel      309 non-null    object 
 5   DTI          309 non-null    int64  
 6   FICO         309 non-null    int64  
 7   HighBalFlag  309 non-null    object 
 8   LockType     309 non-null    object 
 9   LTV          309 non-null    float64
 10  PropOcc      309 non-null    object 
 11  PropState    309 non-null    object 
 12  PropType     309 non-null    object 
 13  PropUnits    309 non-null    int64  
 14  Purpose      309 non-null    object 
 15  RefiType     309 non-null    object 
 16  file_name    309 non-null    object 
dtypes: float64(1), int64(5), object(11)
memory usage: 41.2+ KB


### Loan Results Data-1

In [3]:
# Read multiple json files from a directory and convert them to pandas dataframe
# Add file names as a new column in DataFrame
path_to_json = '/Users/apurwaprasad/Desktop/BlackKnight_data_challenge/Mortgage_Pooling_v1/OptimizationData1/LoanResultsData1/'

data_list = []
json_file_names = []
for file_name in [file for file in os.listdir(path_to_json) if file.endswith('.json')]:
    with open(path_to_json + file_name) as json_file:
        image_name = os.path.basename(file_name)
        data_name = os.path.splitext(image_name)[0]
        data = json.load(json_file)
        data_list.append(data)
        json_file_names.append(data_name)
        
LoanResultsData1 = pd.DataFrame.from_dict(json_normalize(data_list))
LoanResultsData1['file_name'] = json_file_names

LoanResultsData1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 699 entries, 0 to 698
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   cId                   699 non-null    object 
 1   Rank                  699 non-null    int64  
 2   Price                 699 non-null    float64
 3   ValMethodID           699 non-null    object 
 4   ValSpecID             699 non-null    object 
 5   Coupon                699 non-null    float64
 6   DerivedSettleOrdinal  699 non-null    int64  
 7   file_name             699 non-null    object 
dtypes: float64(2), int64(2), object(4)
memory usage: 43.8+ KB


## Optimization Data 2

### Loan Data-2

In [4]:
# Read multiple json files from a directory and convert them to pandas dataframe
# Add file names as a new column in DataFrame
path_to_json = '/Users/apurwaprasad/Desktop/BlackKnight_data_challenge/Mortgage_Pooling_v1/OptimizationData2/LoanData2/'
data_list = []
json_file_names = []
for file_name in [file for file in os.listdir(path_to_json) if file.endswith('.json')]:
    with open(path_to_json + file_name) as json_file:
        image_name = os.path.basename(file_name)
        data_name = os.path.splitext(image_name)[0]
        data = json.load(json_file)
        data_list.append(data)
        json_file_names.append(data_name)
        
LoanData2 = pd.DataFrame.from_dict(json_normalize(data_list))
LoanData2['file_name'] = json_file_names

LoanData2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3556 entries, 0 to 3555
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   AmortTerm    3556 non-null   int64  
 1   Amount       3556 non-null   float64
 2   cCode        3556 non-null   int64  
 3   Channel      3556 non-null   object 
 4   cID          3556 non-null   int64  
 5   DTI          3556 non-null   float64
 6   FICO         3556 non-null   int64  
 7   HighBalFlag  3556 non-null   object 
 8   LTV          3556 non-null   float64
 9   PropOcc      3556 non-null   object 
 10  PropState    3556 non-null   object 
 11  PropType     3556 non-null   object 
 12  PropUnits    3556 non-null   int64  
 13  Purpose      3556 non-null   object 
 14  RefiType     3556 non-null   object 
 15  file_name    3556 non-null   object 
dtypes: float64(3), int64(5), object(8)
memory usage: 444.6+ KB


### Loan Results Data-2

In [5]:
# Read multiple json files from a directory and convert them to pandas dataframe
# Add file names as a new column in DataFrame
path_to_json = '/Users/apurwaprasad/Desktop/BlackKnight_data_challenge/Mortgage_Pooling_v1/OptimizationData2/LoanResultsData2/'

data_list = []
json_file_names = []
for file_name in [file for file in os.listdir(path_to_json) if file.endswith('.json')]:
    with open(path_to_json + file_name) as json_file:
        image_name = os.path.basename(file_name)
        data_name = os.path.splitext(image_name)[0]
        data = json.load(json_file)
        data_list.append(data)
        json_file_names.append(data_name)
        
LoanResultsData2 = pd.DataFrame.from_dict(json_normalize(data_list))
LoanResultsData2['file_name'] = json_file_names

LoanResultsData2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44070 entries, 0 to 44069
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   cID                   44070 non-null  int64  
 1   Rank                  44070 non-null  int64  
 2   Coupon                44070 non-null  float64
 3   DerivedSettleOrdinal  44070 non-null  int64  
 4   Price                 44070 non-null  float64
 5   ValMethodID           44070 non-null  object 
 6   ValSpecID             44070 non-null  object 
 7   Pool Type             44070 non-null  object 
 8   Servicing Retention   44070 non-null  object 
 9   Pool Balance Type     44070 non-null  object 
 10  Agency                44070 non-null  object 
 11  file_name             44070 non-null  object 
dtypes: float64(2), int64(3), object(7)
memory usage: 4.0+ MB


## Optimization Data 3

### Loan Data- 3

In [6]:
# Read multiple json files from a directory and convert them to pandas dataframe
# Add file names as a new column in DataFrame
path_to_json = '/Users/apurwaprasad/Desktop/BlackKnight_data_challenge/Mortgage_Pooling_v1/OptimizationData3/LoanData3/'

data_list = []
json_file_names = []
for file_name in [file for file in os.listdir(path_to_json) if file.endswith('.json')]:
    with open(path_to_json + file_name) as json_file:
        image_name = os.path.basename(file_name)
        data_name = os.path.splitext(image_name)[0]
        data = json.load(json_file)
        data_list.append(data)
        json_file_names.append(data_name)
        
LoanData3 = pd.DataFrame.from_dict(json_normalize(data_list))
LoanData3['file_name'] = json_file_names

LoanData3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3556 entries, 0 to 3555
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   AmortTerm    3556 non-null   int64  
 1   Amount       3556 non-null   float64
 2   cCode        3556 non-null   int64  
 3   Channel      3556 non-null   object 
 4   cID          3556 non-null   int64  
 5   DTI          3556 non-null   float64
 6   FICO         3556 non-null   int64  
 7   HighBalFlag  3556 non-null   object 
 8   LTV          3556 non-null   float64
 9   PropOcc      3556 non-null   object 
 10  PropState    3556 non-null   object 
 11  PropType     3556 non-null   object 
 12  PropUnits    3556 non-null   int64  
 13  Purpose      3556 non-null   object 
 14  RefiType     3556 non-null   object 
 15  file_name    3556 non-null   object 
dtypes: float64(3), int64(5), object(8)
memory usage: 444.6+ KB


### Loan Results Data-3

In [7]:
# Read multiple json files from a directory and convert them to pandas dataframe
# Add file names as a new column in DataFrame
path_to_json = '/Users/apurwaprasad/Desktop/BlackKnight_data_challenge/Mortgage_Pooling_v1/OptimizationData3/LoanResultsData3/'

data_list = []
json_file_names = []
for file_name in [file for file in os.listdir(path_to_json) if file.endswith('.json')]:
    with open(path_to_json + file_name) as json_file:
        image_name = os.path.basename(file_name)
        data_name = os.path.splitext(image_name)[0]
        data = json.load(json_file)
        data_list.append(data)
        json_file_names.append(data_name)
        
LoanResultsData3 = pd.DataFrame.from_dict(json_normalize(data_list))
LoanResultsData3['file_name'] = json_file_names

LoanResultsData3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44070 entries, 0 to 44069
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   cID                   44070 non-null  int64  
 1   Rank                  44070 non-null  int64  
 2   Coupon                44070 non-null  float64
 3   DerivedSettleOrdinal  44070 non-null  int64  
 4   Price                 44070 non-null  float64
 5   ValMethodID           44070 non-null  object 
 6   ValSpecID             44070 non-null  object 
 7   Pool Type             44070 non-null  object 
 8   Servicing Retention   44070 non-null  object 
 9   Pool Balance Type     44070 non-null  object 
 10  Agency                44070 non-null  object 
 11  Servicer              44070 non-null  object 
 12  file_name             44070 non-null  object 
dtypes: float64(2), int64(3), object(8)
memory usage: 4.4+ MB


## PoolOptimization Data for TC v5

### BaselineConstraintsSetB

In [8]:
# Read multiple json files from a directory and convert them to pandas dataframe
# Add file names as a new column in DataFrame
path_to_json = '/Users/apurwaprasad/Desktop/BlackKnight_data_challenge/Mortgage_Pooling_v1/PoolOptimizationDataforTCv5/BaselineConstraintsSetB/'

data_list = []
json_file_names = []
for file_name in [file for file in os.listdir(path_to_json) if file.endswith('.json')]:
    with open(path_to_json + file_name) as json_file:
        image_name = os.path.basename(file_name)
        data_name = os.path.splitext(image_name)[0]
        data = json.load(json_file)
        data_list.append(data)
        json_file_names.append(data_name)
        
BaselineConstraintsSetB = pd.DataFrame.from_dict(json_normalize(data_list))
BaselineConstraintsSetB['file_name'] = json_file_names

BaselineConstraintsSetB.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3549 entries, 0 to 3548
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Loan ID           3549 non-null   int64  
 1   Price             3549 non-null   float64
 2   Selected Pool ID  3549 non-null   object 
 3   Servicer          3549 non-null   object 
 4   file_name         3549 non-null   object 
dtypes: float64(1), int64(1), object(3)
memory usage: 138.8+ KB


### EligiblePricingCombinations

In [9]:
# Read multiple json files from a directory and convert them to pandas dataframe
# Add file names as a new column in DataFrame
path_to_json = '/Users/apurwaprasad/Desktop/BlackKnight_data_challenge/Mortgage_Pooling_v1/PoolOptimizationDataforTCv5/EligiblePricingCombinations/'

data_list = []
json_file_names = []
for file_name in [file for file in os.listdir(path_to_json) if file.endswith('.json')]:
    with open(path_to_json + file_name) as json_file:
        image_name = os.path.basename(file_name)
        data_name = os.path.splitext(image_name)[0]
        data = json.load(json_file)
        data_list.append(data)
        json_file_names.append(data_name)
        
EligiblePricingCombinations = pd.DataFrame.from_dict(json_normalize(data_list))
EligiblePricingCombinations['file_name'] = json_file_names

EligiblePricingCombinations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45913 entries, 0 to 45912
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   LoanID         45913 non-null  int64  
 1   Price, P_ijk   45913 non-null  float64
 2   Pool Opton, j  45913 non-null  object 
 3   Servicer, k    45913 non-null  object 
 4   file_name      45913 non-null  object 
dtypes: float64(1), int64(1), object(3)
memory usage: 1.8+ MB


### LoanData

In [10]:
# Read multiple json files from a directory and convert them to pandas dataframe
# Add file names as a new column in DataFrame
path_to_json = '/Users/apurwaprasad/Desktop/BlackKnight_data_challenge/Mortgage_Pooling_v1/PoolOptimizationDataforTCv5/LoanData/'

data_list = []
json_file_names = []
for file_name in [file for file in os.listdir(path_to_json) if file.endswith('.json')]:
    with open(path_to_json + file_name) as json_file:
        image_name = os.path.basename(file_name)
        data_name = os.path.splitext(image_name)[0]
        data = json.load(json_file)
        data_list.append(data)
        json_file_names.append(data_name)
        
LoanData = pd.DataFrame.from_dict(json_normalize(data_list))
LoanData['file_name'] = json_file_names

LoanData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3549 entries, 0 to 3548
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   LoanID       3549 non-null   int64  
 1   Amount       3549 non-null   float64
 2   FICO         3549 non-null   int64  
 3   DTI          3549 non-null   float64
 4   HighBalFlag  3549 non-null   object 
 5   PropOcc      3549 non-null   object 
 6   PropState    3549 non-null   object 
 7   PropType     3549 non-null   object 
 8   Purpose      3549 non-null   object 
 9   file_name    3549 non-null   object 
dtypes: float64(2), int64(2), object(6)
memory usage: 277.4+ KB


### PoolOptionData

In [11]:
# Read multiple json files from a directory and convert them to pandas dataframe
# Add file names as a new column in DataFrame
path_to_json = '/Users/apurwaprasad/Desktop/BlackKnight_data_challenge/Mortgage_Pooling_v1/PoolOptimizationDataforTCv5/PoolOptionData/'

data_list = []
json_file_names = []
for file_name in [file for file in os.listdir(path_to_json) if file.endswith('.json')]:
    with open(path_to_json + file_name) as json_file:
        image_name = os.path.basename(file_name)
        data_name = os.path.splitext(image_name)[0]
        data = json.load(json_file)
        data_list.append(data)
        json_file_names.append(data_name)
        
PoolOptionData = pd.DataFrame.from_dict(json_normalize(data_list))
PoolOptionData['file_name'] = json_file_names

PoolOptionData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Pool Option, j     458 non-null    object
 1   Pool Type          458 non-null    object
 2   Pool Balance Type  458 non-null    object
 3   Agency             458 non-null    object
 4   Servicer           458 non-null    object
 5   file_name          458 non-null    object
dtypes: object(6)
memory usage: 21.6+ KB


# Summarising the extracted data

# Loan Data:
LoanData1
RangeIndex: 309 entries, 0 to 308
Data columns (total 17 columns)
Extra Column: LockType(Singlevalue-M)

LoanData2
RangeIndex: 3556 entries, 0 to 3555
Data columns (total 16 columns)

LoanData3
RangeIndex: 3556 entries, 0 to 3555
Data columns (total 16 columns)

In [12]:
LoanData1.head()

cID  AmortTerm  Amount  cCode       Channel  DTI  FICO HighBalFlag  \
0  658-0198394        360  379200  C30FN        Retail   18   796           N   
1  685-0203626        360  446300  C30FN  Consumer Dir   32   726           N   
2  655-0199764        360  217500  C30FN        Retail   49   749           N   
3  657-0198422        360  154800  C30FN        Retail   23   770           N   
4  685-0202871        360  553344  C30FN  Consumer Dir   25   798           Y   

  LockType     LTV  PropOcc PropState PropType  PropUnits    Purpose  \
0        M  72.923  Primary        AL      SFR          1  Refinance   
1        M  85.992  Primary        PA      SFR          1  Refinance   
2        M  79.963   Second        AL      SFR          1  Refinance   
3        M  90.000  Primary        GA      SFR          1   Purchase   
4        M  74.776  Primary        CO      SFR          1  Refinance   

   RefiType                             file_name  
0  RateTerm  11a6cdbf-9a50-4f4d-b928-ff4c201234e2  
1  RateTerm  33c5b92b-e185-4aab-9e26-c9d24ffe9f2c  
2  RateTerm  e1e71b28-23e2-4c56-89e6-de9b78d3de1e  
3         N  04e08abb-62c7-4932-a63c-287a5c307d09  
4  RateTerm  1b73ca3b-a857-4d78-9c2f-d64f20a288f2

In [13]:
LoanData2.head()

AmortTerm    Amount    cCode Channel         cID     DTI  FICO HighBalFlag  \
0        180  292000.0   101115  Retail   103472239  47.241   756           N   
1        180  204000.0   101115  Retail   103467486  44.985   774           N   
2        120  240000.0  1011106  Retail   103445492  45.539   795           N   
3        360  686882.0   102130  iMort2  3800031495  32.542   779           Y   
4        180  399800.0   101115  Retail   103497584  32.635   767           N   

      LTV  PropOcc PropState PropType  PropUnits    Purpose   RefiType  \
0  50.258  Primary        GA   SFR_At          1  Refinance    Cashout   
1  80.000  Primary        IN   SFR_Dt          1  Refinance    Cashout   
2  67.797  Primary        TX   SFR_At          1  Refinance    Cashout   
3  95.000  Primary        CA   SFR_At          1   Purchase   Purchase   
4  69.530  Primary        MN   SFR_At          1  Refinance  Rate/Term   

                              file_name  
0  6616f717-9a2a-4512-8711-0bd4c4ba11f4  
1  c0d174af-06a4-4392-879d-6d11d28d0e56  
2  e8dc2db2-d050-4575-9206-e006afef1526  
3  cff26188-ff85-45e7-82ba-d769867a29e8  
4  85a586f4-d7bf-4ec0-aa00-1f9b7dc5b893

In [14]:
LoanData3.head()

AmortTerm    Amount   cCode Channel         cID     DTI  FICO HighBalFlag  \
0        360  255000.0  101130  Retail   103513473  44.400   795           N   
1        360  640000.0  101130  Retail   600141458  43.406   700           Y   
2        360  452000.0  101130  iMort2  3000582053  48.455   683           N   
3        180  200000.0  101115  Retail   103472554  20.237   792           N   
4        360  336000.0  101130  Retail   600146360  48.505   765           N   

      LTV  PropOcc PropState PropType  PropUnits    Purpose   RefiType  \
0  85.000      NOO        CA  CondoLo          1   Purchase   Purchase   
1  71.508  Primary        HI   SFR_Dt          1  Refinance    Cashout   
2  72.089  Primary        WA   SFR_At          1  Refinance  Rate/Term   
3  79.051  Primary        PA   SFR_Dt          1  Refinance    Cashout   
4  80.000  Primary        NJ   SFR_Dt          1   Purchase   Purchase   

                              file_name  
0  d56f3320-caf3-4baa-98ec-badc116a0d92  
1  8a556146-d407-4bbd-980b-cc4fc9245ead  
2  7d08bd39-686f-4d16-9dc7-1af4566a997c  
3  795376de-1d76-4b1f-a77b-440079f8cfd6  
4  ef312cc0-b8b1-4ab7-8f90-8b1cf6f63e14

# Loan Results Data :
LoanResultsData1
RangeIndex: 699 entries, 0 to 698
Data columns (total 8 columns)

LoanResultsData2
RangeIndex: 44070 entries, 0 to 44069
Data columns (total 12 columns)

LoanResultsData3
RangeIndex: 44070 entries, 0 to 44069
Data columns (total 13 columns)

In [15]:
LoanResultsData1.head()

cId  Rank       Price    ValMethodID ValSpecID  Coupon  \
0  685-0201519     1  100.757217    Freddie_MBS      FG30     2.5   
1  604-0199765     5  102.817552  FHLMC_NewProd      FG30     3.0   
2  610-0198729    11  103.659681       FNMA_175      FN30     3.5   
3  685-0202407     1  105.786204       FNMA_HLB      FN30     4.5   
4  605-0204204     6  102.155073  FHLMC_NewProd      FG30     4.5   

   DerivedSettleOrdinal                             file_name  
0                201911  278a721f-f011-4274-a9e8-4998ddff9fcc  
1                201911  f18cf9f0-99d1-403d-aa8c-dee30cd568f0  
2                201911  3416fb03-9522-4e5a-815e-dd35106157bb  
3                201911  a4ea48e8-7629-483a-acc9-266effa7cad2  
4                201911  fc906f5b-7c19-4b6a-9230-80c09f195f3a

In [16]:
LoanResultsData2.head()

cID  Rank  Coupon  DerivedSettleOrdinal       Price  \
0   600146814    18     4.0                201912  101.943561   
1   103409559     5     2.0                201912  100.517589   
2   103475407     9     3.0                201912   98.549534   
3  3000519031     2     3.5                202001  103.042494   
4   103446434     8     2.5                201912  100.425097   

            ValMethodID ValSpecID      Pool Type Servicing Retention  \
0  SP_TWO_FN30_175k_COI      FN30  Single-Issuer            Co-Issue   
1       FN_MBS Retained      FN15   Multi-Issuer            Retained   
2       FN_MBS Retained      FN30   Multi-Issuer            Retained   
3    Pingora_FN_MBS_COI      FN30   Multi-Issuer            Co-Issue   
4        Two_FN_MBS_COI      FN30   Multi-Issuer            Co-Issue   

  Pool Balance Type      Agency                             file_name  
0  Standard Balance  Fannie Mae  037d37ba-9cf4-4124-99de-54fac1e25c98  
1  Standard Balance  Fannie Mae  f9d5c443-d65a-4833-aa06-5b79d1ae75c6  
2  Standard Balance  Fannie Mae  2825d88e-1b2e-4c34-bed3-0de3582d2c72  
3  Standard Balance  Fannie Mae  69a9dba1-8b3b-4496-bde4-3cbbae94e225  
4  Standard Balance  Fannie Mae  a52a0455-54e8-4576-829f-2482f49e6a73

In [17]:
LoanResultsData3.head()

cID  Rank  Coupon  DerivedSettleOrdinal       Price   ValMethodID  \
0   103497067    12     3.0                201912  100.502224        FN_MBS   
1   103449385     7     2.0                201912  100.538767        FN_MBS   
2   103495335     6     3.5                201912   98.806753        FN_MBS   
3   103392396     7     3.0                201912  103.039599        FG_MBS   
4  3000593179     3     3.0                201912  103.049408  SP_FG30_175k   

  ValSpecID      Pool Type Servicing Retention Pool Balance Type       Agency  \
0      FN30   Multi-Issuer            Retained  Standard Balance   Fannie Mae   
1      FN15   Multi-Issuer            Retained  Standard Balance   Fannie Mae   
2      FN30   Multi-Issuer            Co-Issue  Standard Balance   Fannie Mae   
3      FG15   Multi-Issuer            Retained  Standard Balance  Freddie Mac   
4      FG30  Single-Issuer            Co-Issue  Standard Balance  Freddie Mac   

      Servicer                             file_name  
0     Retained  6cdd3d1d-db5a-4ed4-9356-97dbf2a372bc  
1     Retained  bff8f1a5-13b8-4554-8f71-aa1814098ca0  
2  Two Harbors  81d3bd2a-e80b-44ac-bc45-38ff397de3b7  
3     Retained  f1d3432f-08b3-4e8b-aef8-f7822e7a274a  
4      Pingora  7d4843bd-1bd8-4f00-a54c-e27e02f65896

# Pool Optimization Data for TCv5
BaselineConstraintsSetB
RangeIndex: 3549 entries, 0 to 3548
Data columns (total 5 columns)


EligiblePricingCombinations
RangeIndex: 45913 entries, 0 to 45912
Data columns (total 5 columns)

LoanData
RangeIndex: 3549 entries, 0 to 3548
Data columns (total 10 columns)

PoolOptionData
RangeIndex: 458 entries, 0 to 457
Data columns (total 6 columns)

In [18]:
BaselineConstraintsSetB.head()

Loan ID       Price Selected Pool ID     Servicer  \
0   600139104  103.959696         pool_116     Retained   
1   103471260  101.254270           pool_7      Pingora   
2   103476643  101.255380           pool_6  Two Harbors   
3   103468867  101.110244           pool_6  Two Harbors   
4  3200065197  101.613143          pool_13     Retained   

                              file_name  
0  07a0c0e1-ad33-4106-b90a-b9e145122f35  
1  87fa433f-6d95-4191-9698-36f505228ca7  
2  151a53a6-d888-42c3-b201-4b29b62a0979  
3  c4631f71-72a5-4ec1-8477-8ecbbb935650  
4  29224da3-f964-4b49-882f-f3b0b15fb99f

In [19]:
EligiblePricingCombinations.head()

LoanID  Price, P_ijk Pool Opton, j  Servicer, k  \
0   600143433    102.391367      pool_232     Retained   
1   103490174    100.473387      pool_231      Pingora   
2   103476469    100.479324      pool_233  Two Harbors   
3   103443660    102.136004        pool_6  Two Harbors   
4  3000575529    102.976677      pool_249      Pingora   

                              file_name  
0  03b09176-9ddd-4bd9-8df8-4e6b15a611d5  
1  23fc02f0-594e-4624-8201-98cbbaa48abb  
2  a38004bb-e727-449d-9c25-85954dc09b31  
3  accd47eb-7892-4009-8b26-78a2b456487e  
4  6fd6bba3-dc08-4927-8482-19ab6d134bff

In [20]:
LoanData.head()

LoanID    Amount  FICO     DTI HighBalFlag  PropOcc PropState PropType  \
0   103423456  105000.0   685  47.642           N  Primary        WV   SFR_Dt   
1   103480471  301000.0   739  37.421           N  Primary        WA   SFR_At   
2  3000601717  188000.0   802  24.358           N  Primary        CA   SFR_Dt   
3   103479903  326350.0   692  46.753           N  Primary        WA   SFR_Dt   
4   103461196  403000.0   699  49.677           N  Primary        CA   SFR_At   

     Purpose                             file_name  
0    Cashout  1f6acc51-29a8-406c-bb76-7ea154975100  
1  Rate/Term  8d5b53ae-44af-40fa-ace6-589a8864e766  
2  Rate/Term  af2c8882-712c-4a7d-80c3-f1ed6b78e8fe  
3    Cashout  34dc297b-1903-4b7a-92b8-ffba67ff4c4d  
4  Rate/Term  8dbac84d-840d-4236-8ddb-ffbb046269bf

In [21]:
PoolOptionData.head()

Pool Option, j      Pool Type Pool Balance Type       Agency     Servicer  \
0        pool_37   Multi-Issuer      High Balance  Freddie Mac  Two Harbors   
1        pool_43   Multi-Issuer      High Balance  Freddie Mac      Pingora   
2        pool_87  Single-Issuer  Standard Balance  Freddie Mac      Pingora   
3       pool_371  Single-Issuer  Standard Balance   Fannie Mae  Two Harbors   
4        pool_26   Multi-Issuer  Standard Balance  Freddie Mac     Retained   

                              file_name  
0  9a4b0f41-f7f2-43f7-97fb-05e6415292c2  
1  f24dafd6-67d1-4b3e-a655-bea820e1f2cf  
2  f14cab46-795a-4c3b-bbfa-6f207a964494  
3  6f7d8e0b-a8ea-4935-94cb-7439cc164ffe  
4  b83e4c6e-05a9-406b-8e1f-ffca8bf8ca46

# Transforming Data for making visualizations

## LoanData1 and LoanResultsData1
Merging LoanData and LoanResultsData to create one dataframe which can be converted to excel file later.
Using rank column to filter out the cId's with rank 1 to be used and merged with LoanData to find insights further.

In [22]:
LoanResultsData1.count()

cId                     699
Rank                    699
Price                   699
ValMethodID             699
ValSpecID               699
Coupon                  699
DerivedSettleOrdinal    699
file_name               699
dtype: int64

In [23]:
#LoanResultsData1
#Using LoanResultsData1 to identify cId with rank 1
LoanResultsData1_rank1= LoanResultsData1[LoanResultsData1['Rank']==1]
LoanResultsData1_rank1.count()

cId                     157
Rank                    157
Price                   157
ValMethodID             157
ValSpecID               157
Coupon                  157
DerivedSettleOrdinal    157
file_name               157
dtype: int64

In [24]:
#Checking number of duplicates against each cId
LoanResultsData1_rank1['cId'].value_counts()

#three records with same cId

2029-0198814    2
2028-0202874    2
2028-0204027    2
685-0200121     1
685-0201570     1
610-0206086     1
685-0200439     1
685-0205081     1
685-0203703     1
610-0204812     1
685-0202561     1
685-0200607     1
685-0203237     1
685-0198708     1
685-0202608     1
685-0202918     1
655-0198863     1
685-0198723     1
685-0202878     1
685-0202905     1
685-0202254     1
685-0202480     1
685-0202570     1
685-0203143     1
651-0203520     1
685-0201238     1
655-0202593     1
610-0197639     1
652-0201309     1
685-0201743     1
685-0199756     1
685-0202680     1
685-0202501     1
685-0202994     1
685-0203287     1
685-0203044     1
685-0203291     1
685-0202908     1
604-0203714     1
685-0201771     1
685-0202808     1
685-0202150     1
685-0203047     1
604-0201239     1
685-0202383     1
685-0202096     1
604-0206097     1
2032-0204490    1
685-0202297     1
610-0205810     1
602-0205224     1
604-0205323     1
605-0204204     1
685-0201769     1
604-0192297     1
685-020160

In [25]:
#count of all columns
LoanResultsData1.count()

cId                     699
Rank                    699
Price                   699
ValMethodID             699
ValSpecID               699
Coupon                  699
DerivedSettleOrdinal    699
file_name               699
dtype: int64

In [26]:
#count of cId with rank 1
LoanResultsData1_rank1.count()

cId                     157
Rank                    157
Price                   157
ValMethodID             157
ValSpecID               157
Coupon                  157
DerivedSettleOrdinal    157
file_name               157
dtype: int64

In [27]:
#droping duplicate cId's and keeping first record
LoanResultsData1_rank1 = LoanResultsData1_rank1.drop_duplicates(subset='cId', keep="first")
LoanResultsData1_rank1.head()

cId  Rank       Price    ValMethodID ValSpecID  Coupon  \
0    685-0201519     1  100.757217    Freddie_MBS      FG30     2.5   
3    685-0202407     1  105.786204       FNMA_HLB      FN30     4.5   
11   685-0201036     1  101.853527  FHLMC_NewProd      FG30     3.0   
14  2028-0202874     1  102.560527  FHLMC_NewProd      FG30     3.0   
18   685-0202570     1  103.273502       FNMA_HLB      FN30     3.0   

    DerivedSettleOrdinal                             file_name  
0                 201911  278a721f-f011-4274-a9e8-4998ddff9fcc  
3                 201911  a4ea48e8-7629-483a-acc9-266effa7cad2  
11                201911  cb129ccb-930d-41fa-9ada-e99f7d6e068d  
14                201911  0a570170-ba8c-42ce-bacb-06b142edee36  
18                201911  133d6182-1cc0-4760-85a5-8f84fb5e4c47

In [28]:
#Verifying records after dropping
LoanResultsData1_rank1.count()

cId                     154
Rank                    154
Price                   154
ValMethodID             154
ValSpecID               154
Coupon                  154
DerivedSettleOrdinal    154
file_name               154
dtype: int64

In [29]:
#LoanData1
#count of all columns
LoanData1.count()

cID            309
AmortTerm      309
Amount         309
cCode          309
Channel        309
DTI            309
FICO           309
HighBalFlag    309
LockType       309
LTV            309
PropOcc        309
PropState      309
PropType       309
PropUnits      309
Purpose        309
RefiType       309
file_name      309
dtype: int64

In [30]:
#Merging both dataframes LoanData1 and LoanResultsData1
Loan1_final = pd.merge(LoanData1,LoanResultsData1_rank1,left_on='cID',right_on='cId', how= 'inner')

#Verifying number of records
Loan1_final.count()

cID                     309
AmortTerm               309
Amount                  309
cCode                   309
Channel                 309
DTI                     309
FICO                    309
HighBalFlag             309
LockType                309
LTV                     309
PropOcc                 309
PropState               309
PropType                309
PropUnits               309
Purpose                 309
RefiType                309
file_name_x             309
cId                     309
Rank                    309
Price                   309
ValMethodID             309
ValSpecID               309
Coupon                  309
DerivedSettleOrdinal    309
file_name_y             309
dtype: int64

## LoanData2 and LoanResultsData2
Merging LoanData2 and LoanResultsData2 to create one dataframe which can be converted to excel file later.
Using rank column to filter out the cId's with rank 1 to be used and merged with LoanData to find insights further.

### LoanResultsData2

In [31]:
#LoanResultsData2
#Using LoanResultsData2 to identify cID with rank 1
LoanResultsData2_rank1= LoanResultsData2[LoanResultsData2['Rank']==1]
LoanResultsData2_rank1.count()

cID                     3556
Rank                    3556
Coupon                  3556
DerivedSettleOrdinal    3556
Price                   3556
ValMethodID             3556
ValSpecID               3556
Pool Type               3556
Servicing Retention     3556
Pool Balance Type       3556
Agency                  3556
file_name               3556
dtype: int64

In [57]:
#Checking number of duplicates against each cId
LoanResultsData2_rank1['cID'].value_counts()

#No duplicates found

cID                     3556
Rank                    3556
Coupon                  3556
DerivedSettleOrdinal    3556
Price                   3556
ValMethodID             3556
ValSpecID               3556
Pool Type               3556
Servicing Retention     3556
Pool Balance Type       3556
Agency                  3556
file_name               3556
dtype: int64

In [33]:
LoanResultsData2_rank1.count()

cID                     3556
Rank                    3556
Coupon                  3556
DerivedSettleOrdinal    3556
Price                   3556
ValMethodID             3556
ValSpecID               3556
Pool Type               3556
Servicing Retention     3556
Pool Balance Type       3556
Agency                  3556
file_name               3556
dtype: int64

In [34]:
LoanData2.count()

AmortTerm      3556
Amount         3556
cCode          3556
Channel        3556
cID            3556
DTI            3556
FICO           3556
HighBalFlag    3556
LTV            3556
PropOcc        3556
PropState      3556
PropType       3556
PropUnits      3556
Purpose        3556
RefiType       3556
file_name      3556
dtype: int64

In [35]:
#Merging both dataframes LoanData1 and LoanResultsData1
Loan2_final = pd.merge(LoanData2,LoanResultsData2_rank1,left_on='cID',right_on='cID', how= 'inner')

In [36]:
#Verifying number of records
Loan2_final.count()

AmortTerm               3556
Amount                  3556
cCode                   3556
Channel                 3556
cID                     3556
DTI                     3556
FICO                    3556
HighBalFlag             3556
LTV                     3556
PropOcc                 3556
PropState               3556
PropType                3556
PropUnits               3556
Purpose                 3556
RefiType                3556
file_name_x             3556
Rank                    3556
Coupon                  3556
DerivedSettleOrdinal    3556
Price                   3556
ValMethodID             3556
ValSpecID               3556
Pool Type               3556
Servicing Retention     3556
Pool Balance Type       3556
Agency                  3556
file_name_y             3556
dtype: int64

### LoanResultsData3

## LoanData3 and LoanResultsData3
Merging LoanData3 and LoanResultsData3 to create one dataframe which can be converted to excel file later.
Using rank column to filter out the cId's with rank 1 to be used and merged with LoanData to find insights further.

In [37]:
LoanResultsData3.count()

cID                     44070
Rank                    44070
Coupon                  44070
DerivedSettleOrdinal    44070
Price                   44070
ValMethodID             44070
ValSpecID               44070
Pool Type               44070
Servicing Retention     44070
Pool Balance Type       44070
Agency                  44070
Servicer                44070
file_name               44070
dtype: int64

In [38]:
#Using LoanResultsData3 to identify cId with rank 1
LoanResultsData3_rank1= LoanResultsData3[LoanResultsData3['Rank']==1]
LoanResultsData3_rank1.count()

cID                     3556
Rank                    3556
Coupon                  3556
DerivedSettleOrdinal    3556
Price                   3556
ValMethodID             3556
ValSpecID               3556
Pool Type               3556
Servicing Retention     3556
Pool Balance Type       3556
Agency                  3556
Servicer                3556
file_name               3556
dtype: int64

In [39]:
#Checking number of duplicates against each cId
LoanResultsData3_rank1['cID'].value_counts()

#No duplicates found

600139775     1
3200056018    1
600144618     1
3000574697    1
600142567     1
103457482     1
103455730     1
3200060128    1
600146653     1
600144602     1
103464189     1
103444125     1
600134355     1
103473869     1
3000576752    1
103455434     1
600146633     1
103488198     1
3000593088    1
3000578751    1
103451324     1
103487354     1
103469751     1
103471798     1
103430834     1
3000584990    1
103488238     1
103457521     1
103484076     1
103492384     1
103488314     1
103476024     1
103482161     1
103447342     1
103421508     1
3000596625    1
103471914     1
103449385     1
103476008     1
600146725     1
103473953     1
103461663     1
103506677     1
103457565     1
3000591129    1
103475988     1
3000566545    1
103448055     1
103439110     1
103447272     1
3200064258    1
103463517     1
600129732     1
3200058105    1
3000584880    1
600144554     1
600144970     1
103432765     1
103488090     1
103465557     1
3200054781    1
600146513     1
60014446

In [40]:
LoanData3.count()

AmortTerm      3556
Amount         3556
cCode          3556
Channel        3556
cID            3556
DTI            3556
FICO           3556
HighBalFlag    3556
LTV            3556
PropOcc        3556
PropState      3556
PropType       3556
PropUnits      3556
Purpose        3556
RefiType       3556
file_name      3556
dtype: int64

In [41]:
#Merging both dataframes LoanData1 and LoanResultsData1
Loan3_final = pd.merge(LoanData3,LoanResultsData3_rank1,left_on='cID',right_on='cID', how= 'inner')

#Verifying number of records
Loan3_final.count()

AmortTerm               3556
Amount                  3556
cCode                   3556
Channel                 3556
cID                     3556
DTI                     3556
FICO                    3556
HighBalFlag             3556
LTV                     3556
PropOcc                 3556
PropState               3556
PropType                3556
PropUnits               3556
Purpose                 3556
RefiType                3556
file_name_x             3556
Rank                    3556
Coupon                  3556
DerivedSettleOrdinal    3556
Price                   3556
ValMethodID             3556
ValSpecID               3556
Pool Type               3556
Servicing Retention     3556
Pool Balance Type       3556
Agency                  3556
Servicer                3556
file_name_y             3556
dtype: int64

In [42]:
#Exporting out data to xlsx for all merged Loan Data's with Loan Data Result's 
Loan1_final.to_excel('/Users/apurwaprasad/Desktop/BlackKnight_data_challenge/Data/Loan1_final.xlsx', index = False)
Loan2_final.to_excel('/Users/apurwaprasad/Desktop/BlackKnight_data_challenge/Data/Loan2_final.xlsx', index = False)
Loan3_final.to_excel('/Users/apurwaprasad/Desktop/BlackKnight_data_challenge/Data/Loan3_final.xlsx', index = False)

## LoanData3 and LoanResultsData3
Merging LoanData3 and LoanResultsData3 to create one dataframe which can be converted to excel file later.
Using rank column to filter out the cId's with rank 1 to be used and merged with LoanData to find insights further.

In [43]:
EligiblePricingCombinations.head()

LoanID  Price, P_ijk Pool Opton, j  Servicer, k  \
0   600143433    102.391367      pool_232     Retained   
1   103490174    100.473387      pool_231      Pingora   
2   103476469    100.479324      pool_233  Two Harbors   
3   103443660    102.136004        pool_6  Two Harbors   
4  3000575529    102.976677      pool_249      Pingora   

                              file_name  
0  03b09176-9ddd-4bd9-8df8-4e6b15a611d5  
1  23fc02f0-594e-4624-8201-98cbbaa48abb  
2  a38004bb-e727-449d-9c25-85954dc09b31  
3  accd47eb-7892-4009-8b26-78a2b456487e  
4  6fd6bba3-dc08-4927-8482-19ab6d134bff

In [44]:
#Using the LoanID to find max Price, P_ijk and saving one LoanId with max price
EligiblePricingComb = EligiblePricingCombinations.iloc[EligiblePricingCombinations.groupby('LoanID')['Price, P_ijk'
].agg(pd.Series.idxmax)]
EligiblePricingComb=pd.DataFrame(EligiblePricingComb)
EligiblePricingComb.count()

LoanID           3549
Price, P_ijk     3549
Pool Opton, j    3549
Servicer, k      3549
file_name        3549
dtype: int64

In [45]:
PoolOptionData.head()

Pool Option, j      Pool Type Pool Balance Type       Agency     Servicer  \
0        pool_37   Multi-Issuer      High Balance  Freddie Mac  Two Harbors   
1        pool_43   Multi-Issuer      High Balance  Freddie Mac      Pingora   
2        pool_87  Single-Issuer  Standard Balance  Freddie Mac      Pingora   
3       pool_371  Single-Issuer  Standard Balance   Fannie Mae  Two Harbors   
4        pool_26   Multi-Issuer  Standard Balance  Freddie Mac     Retained   

                              file_name  
0  9a4b0f41-f7f2-43f7-97fb-05e6415292c2  
1  f24dafd6-67d1-4b3e-a655-bea820e1f2cf  
2  f14cab46-795a-4c3b-bbfa-6f207a964494  
3  6f7d8e0b-a8ea-4935-94cb-7439cc164ffe  
4  b83e4c6e-05a9-406b-8e1f-ffca8bf8ca46

In [46]:
PoolOptionData.count()

Pool Option, j       458
Pool Type            458
Pool Balance Type    458
Agency               458
Servicer             458
file_name            458
dtype: int64

In [47]:
#Merging EligiblePricingCombinations dataframe with Loan having max price with PoolOptionData.
#EligiblePricingComb1 will give deatils about Loan, Max price issued, pool details
EligiblePricingComb1 = pd.merge(EligiblePricingComb, PoolOptionData,
                                left_on=['Pool Opton, j'],right_on=['Pool Option, j'], how = 'inner')

#EligiblePricingComb1=pd.DataFrame(EligiblePricingComb1)
EligiblePricingComb1.head()

LoanID  Price, P_ijk Pool Opton, j Servicer, k  \
0   102776883    102.009171      pool_249     Pingora   
1   103457579    102.772171      pool_249     Pingora   
2   103496164    103.356714      pool_249     Pingora   
3  3000448519    102.604447      pool_249     Pingora   
4  3000493349    102.969091      pool_249     Pingora   

                            file_name_x Pool Option, j     Pool Type  \
0  043b103a-2aac-4f27-913d-29813a1329b4       pool_249  Multi-Issuer   
1  b2011636-0e8c-4c3f-adb1-3856bd159083       pool_249  Multi-Issuer   
2  7466f06a-b0b6-4675-bdcd-6463232310f3       pool_249  Multi-Issuer   
3  85342ada-79c5-4e1b-86c9-12cb8b442a2e       pool_249  Multi-Issuer   
4  2c551270-6077-4691-863e-8b944c92f197       pool_249  Multi-Issuer   

  Pool Balance Type      Agency Servicer                           file_name_y  
0  Standard Balance  Fannie Mae  Pingora  2c30a808-2055-4d88-9ca2-42afa3caa3e4  
1  Standard Balance  Fannie Mae  Pingora  2c30a808-2055-4d88-9ca2-42afa3caa3e4  
2  Standard Balance  Fannie Mae  Pingora  2c30a808-2055-4d88-9ca2-42afa3caa3e4  
3  Standard Balance  Fannie Mae  Pingora  2c30a808-2055-4d88-9ca2-42afa3caa3e4  
4  Standard Balance  Fannie Mae  Pingora  2c30a808-2055-4d88-9ca2-42afa3caa3e4

In [48]:
#Verifying records after merge
EligiblePricingComb1.count()

LoanID               3549
Price, P_ijk         3549
Pool Opton, j        3549
Servicer, k          3549
file_name_x          3549
Pool Option, j       3549
Pool Type            3549
Pool Balance Type    3549
Agency               3549
Servicer             3549
file_name_y          3549
dtype: int64

In [49]:
#exporting the resukting file into excel
EligiblePricingComb1.to_excel('/Users/apurwaprasad/Desktop/BlackKnight_data_challenge/Data/EligiblePricingComb1.xlsx', index = False)

## EligiblePricingComb and LoanData
Merging EligiblePricingComb dataframe with LoanData to get more details about loan and pool.

In [50]:
LoanData.head()

LoanID    Amount  FICO     DTI HighBalFlag  PropOcc PropState PropType  \
0   103423456  105000.0   685  47.642           N  Primary        WV   SFR_Dt   
1   103480471  301000.0   739  37.421           N  Primary        WA   SFR_At   
2  3000601717  188000.0   802  24.358           N  Primary        CA   SFR_Dt   
3   103479903  326350.0   692  46.753           N  Primary        WA   SFR_Dt   
4   103461196  403000.0   699  49.677           N  Primary        CA   SFR_At   

     Purpose                             file_name  
0    Cashout  1f6acc51-29a8-406c-bb76-7ea154975100  
1  Rate/Term  8d5b53ae-44af-40fa-ace6-589a8864e766  
2  Rate/Term  af2c8882-712c-4a7d-80c3-f1ed6b78e8fe  
3    Cashout  34dc297b-1903-4b7a-92b8-ffba67ff4c4d  
4  Rate/Term  8dbac84d-840d-4236-8ddb-ffbb046269bf

In [51]:
LoanData.count()

LoanID         3549
Amount         3549
FICO           3549
DTI            3549
HighBalFlag    3549
PropOcc        3549
PropState      3549
PropType       3549
Purpose        3549
file_name      3549
dtype: int64

In [52]:
#Merging EligiblePricingComb dataframe with LoanData to get more details about loan and pool.
LoanPoolDetails = pd.merge(EligiblePricingComb, LoanData,on=['LoanID'], how = 'inner')

In [53]:
#verifying number of records after merge.
LoanPoolDetails.count()

LoanID           3549
Price, P_ijk     3549
Pool Opton, j    3549
Servicer, k      3549
file_name_x      3549
Amount           3549
FICO             3549
DTI              3549
HighBalFlag      3549
PropOcc          3549
PropState        3549
PropType         3549
Purpose          3549
file_name_y      3549
dtype: int64

In [58]:
#exporting the resukting file into excel
LoanPoolDetails.to_excel('/Users/apurwaprasad/Desktop/BlackKnight_data_challenge/Data/LoanPoolDetails.xlsx', index = False)